ships/0026-shipwright-managed-push.md

```diff
@@ -0,0 +1,160 @@
+
+- Supporting custom certificates for the container registry. Similar to the scenario above, we can add this to the Build's output section and honor this in our push logic.
+- Vulnerability scanning of the image. Once we own the push operation, we can make this a shipwright managed operation that happens before the push to prevent vulnerable images to appear in the container registry. Today, this can be done by build strategy authors, but again makes the strategy more complex. See [kaniko-trivy](https://github.com/shipwright-io/build/blob/v0.8.0/samples/buildstrategy/kaniko/buildstrategy_kaniko-trivy_cr.yaml).
+- Software Bill Of Materials. Having the image locally at hand, we can analyze the content of the image and put together something like an [SPDX document](https://spdx.dev/resources/learn/) and store that somewhere.
+- Image signing. This is loosely coupled because images can only be signed after they are pushed, but still this is a feature that fits well in container image operations that run after the build strategy steps.
+
+### Goals
+
+- Allow the build strategy author to store the image as a tarball to a file system location provided though a new Shipwright system parameter.
+- Introduce mechanisms in Shipwright to read this tarball and push it to the container registry.
+- Continue to support build strategies that perform the push themselves.
+
+### Non-Goals
+
+- Describe any of the follow-on scenarios listed under [Motivation](#motivation) in any detail. These should be covered in separate ships
+
+## Proposal
+
+Today, most BuildRuns result in pods with two contains that run sequentially:
+
+- Shipwright-defined container that loads the sources (can also be multiple if sources of type HTTP are used in combination with a Git or Bundle source)
+- Build strategy step(s) that builds and pushed the image
+
+Going forward we will still support this, but also offer the following option:
+
+- Shipwright-defined container that loads the sources (can also be multiple if sources of type HTTP are used in combination with a Git or Bundle source)
+- Build strategy step(s) that builds the image and stores it in the file system
+- Shipwright-defined container that pushes the image
+
+### API Changes
+
+We will introduce an additional system parameter: `shp-output-directory`. Like `shp-source-root`, this will contain a hardcoded value backed by an emptyDir volume. A good path would be `/workspace/output-image`. This is an extension point for the future for scenarios where we store the image on a persistent volume instead, for example when a Shipwright Build is embedded into a larger pipeline.
+
+A build strategy that uses this parameter = which references `$(params.shp-output-directory)` in any `buildSteps` `commands`, `args` or `env` will be considered a build strategy which does not itself push the image. Any build strategy that does not use this parameter is assumed to perform the image push operation on its own.
```

Comment:
I'm a bit confused here wrt to when Shipwright will decide if it needs to push the container image.

Does it make sense to expose a new field in the `BuildStrategy` API? For example, this would allow strategy authors to specify if the image is pushed by Shipwright (Default), within the strategy itself (Provided), or no image is pushed (None).

```yaml
spec:
  imagePushStrategy: Default | Provided | None
```

Reply:
Hi @adambkaplan, thanks for reading through it. The currently proposed logic is:

* Iterate all `buildStep`s `args`, `command`, and `env` to check if an if them accesses `$(params.shp-output-directory)` (we have code for this already when we decide if we have to add the environment variable for param values from secrets or configmaps: https://github.com/shipwright-io/build/blob/v0.8.0/pkg/reconciler/buildrun/resources/params.go#L346)
* If any step references the parameter, assume that shipwright must push the image, otherwise not

I also considered a separate flag but see it as duplicate information with the risk that the build strategy author sets it inconsistently to how the strategy steps are implemented.

Reply:
That makes sense - thanks for the clarification. I think this suggestion may help:

```suggestion
Shipwright will push the container image if any build step in the build strategy references the `$(params.shp-output-directory)` parameter in the `commands`, `args`, or `env` spec. Any build strategy that does not use this parameter in this manner is assumed to perform the image push operation on its own.
```

Reply:
Done.